### PR TITLE
PoC: Dev of KPIs for mapeditor

### PR DESCRIPTION
**Pull request by KobusVanRooyen**
Developed KPIs which are now written to the optimized results (esdl) file so that they can be viewed in the mapeditor. The KPIs are split into 2 mains groups:
1. Overall network KPIs
2. KPIs per polygon area drawn in the mapeditor

Note:
- Test case still to be developed 
- The code was manually tested with the grow_workflow (test_end_scenario_sizing.py), which was applied to test_case_small_network_ates_buffer_optional_assets. The latter was modified by adding a polygon area for each heating demand, and 1 polygon area for the 2 heat producers.  This modified input esdl file has not been committed to this work, but it can be found at the following link:

[test_case_small_network_with_ates_with_buffer_all_optional.zip](https://github.com/Nieuwe-Warmte-Nu/rtc-tools-heat-network/files/12608110/test_case_small_network_with_ates_with_buffer_all_optional.zip)

### CODE REVIEW

File: src/rtctools_heat_network/workflows/grow_workflow.py [hunk 1]
```diff
@@ -175,7 +175,8 @@ def read(self):
                 else:
                     total_demand += demand_values
                 max_demand = max(demand_values)
-                self.__heat_demand_bounds[f"{demand}.Heat_demand"] = (0.0, 2 * max_demand)
+                # TODO: check why (0.0, 2.0 * max_demand) caused investment to go to upper bound
+                self.__heat_demand_bounds[f"{demand}.Heat_demand"] = (0.0, max_demand)
```
Comment (jimrojerTNO) [inline src/rtctools_heat_network/workflows/grow_workflow.py:179]: let's revert this change as it is related to how the constraints are setup and make a seperate MR for fixing this.

**Review by jimrojerTNO — CHANGES_REQUESTED** [review]

File: src/rtctools_heat_network/workflows/grow_workflow.py (hunk 1 above)
Comment (KobusVanRooyen) [reply src/rtctools_heat_network/workflows/grow_workflow.py:179]: Agreed. We need this fix before PoC release though

File: src/rtctools_heat_network/workflows/io/write_output.py [hunk 2]
```diff
@@ -212,91 +215,347 @@ def _write_updated_esdl(self, db_profiles=False, optimizer_sim=True):
...
+                                try:
+                                    energy_breakdown[asset_type] = np.sum(
+                                        results[f"{asset_name}.Secondary_heat"]
+                                    )
+                                except KeyError:
+                                    pass
+
+                    # Create KPIs by using applicable costs for the specific asset
+                    area_investment_cost += max(
+                        results[self._asset_investment_cost_map[asset_name]][0], 1.0
```
[the remark this answers is not in the document]
Comment (KobusVanRooyen) [reply src/rtctools_heat_network/workflows/io/write_output.py]: Sorry about this one, silly mistake. This was left in place after I copied your prototyping code. It has been removed.

File: src/rtctools_heat_network/workflows/grow_workflow.py (hunk 1 above)
Comment (KobusVanRooyen) [reply src/rtctools_heat_network/workflows/grow_workflow.py]: Code has been reverted back to the original code

**Review by jimrojerTNO — APPROVED** [review]

**Pull request merged**